### PR TITLE
create models, bloc, service

### PR DESCRIPTION
**Pull request by nhatdaoquang25**

### CODE REVIEW

File: lib/src/blocs/coin_bloc/coin_bloc.dart [hunk 1]
```diff
@@ -0,0 +1,11 @@
+import 'package:flutter_bloc/flutter_bloc.dart';
+import 'package:meta/meta.dart';
+
+part 'coin_event.dart';
+part 'coin_state.dart';
```
Comment (thanhphamtrung) [inline lib/src/blocs/coin_bloc/coin_bloc.dart]: import separately instead of using part

File: lib/src/blocs/coin_bloc/coin_event.dart [hunk 2]
```diff
@@ -0,0 +1,4 @@
+part of 'coin_bloc.dart';
```
Comment (thanhphamtrung) [inline lib/src/blocs/coin_bloc/coin_event.dart]: .

File: lib/src/models/coins.dart [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+        image: json["image"] ?? json["image"]["small"],
+        currentPrice: json["current_price"]["usd"].toDouble(),
+        marketCap: json["market_cap"]["usd"],
+        high24H: json["high_24h"]["usd"].toDouble(),
+        low24H: json["low_24h"]["usd"].toDouble(),
+        priceChange24H: json["price_change_24h"].toDouble(),
+        circulatingSupply: json["circulating_supply"].toDouble(),
+        totalSupply: json["total_supply"].toDouble(),
+        maxSupply: json["max_supply"].toDouble(),
+        description: json["description"]["en"],
```
Comment (thanhphamtrung) [inline lib/src/models/coins.dart]: it will be better and safer if you have null coalescing

**Review by thanhphamtrung — CHANGES_REQUESTED** [review]

**Comment by thanhphamtrung** [comment]
because it have some change in UI, so I think we need to show it before merge

File: lib/src/blocs/coin_bloc/coin_bloc.dart (hunk 1 above)
Comment (nhatdaoquang25) [reply lib/src/blocs/coin_bloc/coin_bloc.dart]: I will fix it

File: lib/src/blocs/coin_bloc/coin_event.dart (hunk 2 above)
Comment (nhatdaoquang25) [reply lib/src/blocs/coin_bloc/coin_event.dart]: I will fix it

File: lib/src/models/coins.dart (hunk 3 above)
Comment (nhatdaoquang25) [reply lib/src/models/coins.dart]: I will fix it

File: lib/src/models/coins.dart [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+    required this.priceChangePercentage24H,
+  });
+
+  factory Coins.fromJson(Map<String, dynamic> json) => Coins(
+        id: json["id"],
+        symbol: json["symbol"],
+        name: json["name"],
+        image: json["image"],
+        currentPrice:
+            json["current_price"] == null ? 0 : json["current_price"].toInt(),
```
Comment (thanhphamtrung) [inline lib/src/models/coins.dart]: what happens if  json["current_price"].toInt() can't be done?

File: lib/src/models/coins.dart [hunk 5]
```diff
@@ -3,11 +3,11 @@ class Coins {
...
-  int? currentPrice;
-  double? high24H;
-  double? low24H;
-  double? priceChange24H;
-  double? priceChangePercentage24H;
+  num? currentPrice;
+  num? high24H;
+  num? low24H;
+  num? priceChange24H;
+  num? priceChangePercentage24H;
```
Comment (thanhphamtrung) [inline lib/src/models/coins.dart]: can these fields be null?

**Review by thanhphamtrung — APPROVED** [review]

**Pull request merged**